bindings/python/htcondor2/_collector.py

```diff
@@ -1 +1,128 @@
+            return
+
+        if isinstance(pool, list):
+            str_list = ", ".join(list)
+            _collector_init(self, self._handle, str_list)
+            return
+
+
+    # FIXME: In version 1, `constraint` could also be an ExprTree.
+    def query(self,
+      ad_type: AdType = AdType.Any,
+      constraint: Optional[str] = None,
+      projection: Optional[List[str]] = None,
+      # The documentation does not match the implementation in version 1.
+      # statistics: Optional[List[str]] = None,
+      statistics: Optional[str] = None,
+    ) -> List[classad.ClassAd]:
+        # str(None) is "None", which is a valid ClassAd expression (a bare
+        # attribute reference), so convert to the empty string, instead.
+        # We don't pass `constraint` through unmodified because we'll want
+        # the str() conversions for all of the other data types we care
+        # about to work anyway, and it's easier to do/handle the conversion
+        # in Python.
+        if constraint is None:
+            constraint = ""
+        if projection is None:
+            projection = []
+
+        # We should consider converting automagically, since it's absurd
+        # that this is the only function in the class which takes an AdType
+        # instead of a DaemonType.
+        if not isinstance(ad_type, AdType):
+            raise TypeError("ad_type must be an htcondor.AdType");
+
+        return _collector_query(self._handle, int(ad_type), str(constraint), projection, statistics, None)
+
+
+    def directQuery(self,
+      daemon_type: DaemonType,
+      name: Optional[str] = None,
+      projection: Optional[List[str]] = None,
+      # The documentation does not match the implementation in version 1.
+      # statistics: Optional[List[str]] = None,
+      statistics: str = None,
+    ) -> classad.ClassAd:
+        daemon_ad = self.locate(daemon_type, name)
+        return _collector_query(self._handle, int(daemon_type), daemon_ad, projection, statistics, None)
+
+
+    _for_location = ["MyAddress", "AddressV1", "CondorVersion", "CondorPlatform", "Name", "Machine"]
+
+
+    def locate(self,
+        daemon_type: DaemonType,
+        name: Optional[str] = None,
+    ) -> classad.ClassAd:
+        ad_type = _ad_type_from_daemon_type(daemon_type)
+        if name is not None:
+            constraint = f'stricmp(Name, "{name}") == 0'
+            return _collector_query(self._handle, int(ad_type), constraint, self._for_location, None, name)[0]
+
+        if self.default is False:
+            constraint = True
+            return self.query(ad_type, constraint, self._for_location)[0]
+
+        return _collector_locate_local(self, self._handle, int(daemon_type))
+
+
+    def locateAll(self,
+        daemon_type: DaemonType,
+    ) -> List[classad.ClassAd]:
+        ad_type = _ad_type_from_daemon_type(daemon_type)
+        return self.query(ad_type, projection=Collector._for_location)
```

Comment:
I notice the projection is set to `Collector._for_location` while earlier uses of the same list (lines 107 & 111) use `self._for_location`. I am just curious if there is a particular reason that this list needs to be referenced differently? And then raise the question of should they be the same and which version should be used?

Reply:
It probably needs to be `Collector._for_location` in both places.  I swear I tested this code.

Reply:
Actually, Python doesn't care, but `Collector._for_location` is clearer (because `_for_location` should be a constant, not a per-object value).